mev-build-rs/src/mempool_builder/builder/payload_builder.rs

```diff
@@ -0,0 +1,271 @@
+    }
+
+    // TODO can abort early if not better fee-wise
+
+    // NOTE: payment code for EOB payment
+    // TODO: refactor this lol
+    let integral_percent = (context.bid_percent * 100.0) as u64;
+    let payment = total_fees * U256::from(integral_percent) / U256::from(100);
+    let revenue = total_fees - payment;
+    let total_payment = context.subsidy + payment;
+
+    let sender = context.builder_wallet.address();
+    let signer_account = db.load_account(sender.into())?;
+    let nonce = signer_account.info.nonce;
+
+    let fee_recipient = ethers_H160::from_slice(context.proposer_fee_recipient.as_ref());
+    let tx = Eip1559TransactionRequest::new()
+        .from(sender)
+        .to(fee_recipient)
+        .gas(21000)
+        .max_fee_per_gas(base_fee)
+        .max_priority_fee_per_gas(0)
+        .value(total_payment)
+        .data(ethers::types::Bytes::default())
+        .access_list(ethers::types::transaction::eip2930::AccessList::default())
+        .nonce(nonce)
+        .chain_id(context.cfg_env.chain_id.to::<u64>());
+    let tx = TypedTransaction::Eip1559(tx);
+    let wallet = &context.builder_wallet;
+    let signature = wallet.sign_transaction_sync(&tx).expect("can make transaction");
+    let tx_encoded = tx.rlp_signed(&signature);
+
+    let payment_tx = TransactionSigned::decode_enveloped(Bytes::from(tx_encoded.as_ref()))
+        .expect("can decode valid txn");
+
+    let payment_tx =
+        TransactionSignedEcRecovered::from_signed_transaction(payment_tx, sender.into());
+
+    let env = Env {
+        cfg: context.cfg_env.clone(),
+        block: context.block_env.clone(),
+        tx: tx_env_with_recovered(&payment_tx),
+    };
+
+    let mut evm = revm::EVM::with_env(env);
+    evm.database(&mut db);
+
+    let ResultAndState { result, state } = match evm.transact() {
+        Ok(result) => result,
+        Err(err) => match err {
+            err => return Err(Error::Execution(err)),
+        },
+    };
+
+    commit_state_changes(&mut db, &mut post_state, block_number, state, true);
+
+    let gas_used = result.gas_used();
+    cumulative_gas_used += gas_used;
+
+    post_state.add_receipt(
+        block_number,
+        Receipt {
+            tx_type: payment_tx.tx_type(),
+            success: result.is_success(),
+            cumulative_gas_used,
+            logs: result.logs().into_iter().map(into_reth_log).collect(),
+        },
+    );
+
+    executed_txs.push(payment_tx.into_signed());
+
+    let withdrawals_root = process_withdrawals(
+        &context.withdrawals,
+        &context.chain_spec,
+        &mut db,
+        &mut post_state,
+        context.timestamp,
+        block_number,
+    )?;
+
+    let receipts_root = post_state.receipts_root(block_number);
+    let logs_bloom = post_state.logs_bloom(block_number);
+    let state_root = db.db.0.state_root(post_state)?;
+    let transactions_root = proofs::calculate_transaction_root(&executed_txs);
+
+    let header = Header {
```

Comment:
@ralexstokes TODO: consider using the `ethereum-consensus` type here to skip the "EL legacy" stuff we don't wanna think about